### PR TITLE
torchdim Python port

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.11.0) (oldest at bottom):
* #163340
* __->__ #160236

The big semantic change (and the reason for this port) is that we no longer monkeypatch Tensor with torchdim's special methods. The new algorithm for handling dispatch is that we first land in `__torch_function__` and we see if a special FCD implementation needs to be dispatch to first, and if there is nothing we fallback to the standard level strategy.

Because there is no longer C binding equivalent of classes, we've condensed _C.Dim and Dim together, and similar for Tensor. This resulted in some bugs as the Python API is sometimes different from the C API. I've attempted to disambiguate these but there may still be mistakes (many early bugs were due to this problem). Dim and DimEntry are especially painful as Dim must abide by Tensor equality semantics, but is pointer equality in C (DimEntry doesn't have this problem). Another difference between C/Python that is subtle is we no longer get implicit conversions from Dim to DimEntry, this also caused some bugs.

Much of the mechanical porting work was done by claude code. I have a separate PR that deletes functorch._C, but it was useful having dim.cpp to point claude at it so I haven't done it in this PR. From a reviewing perspective, I need to re-review that I didn't forget to port anything, some noticeably missing "small" things are patched_dim_method. I am still in progress of carefully doing a side-by-side review of ports; "simplifications" from claude code were also a major source of bugs.

There are two major feature gaps in the implementation:

- DelayedTensor and dot handling are not implemented yet. This should be reasonably easy, just need to do it.  However, for the purposes of sharded propagation it is actually better not to reconstruct matmuls.
- Splitting dimensions with an index like `[x, y]` doesn't work. The problem is that `__getitem__` interprets this as advanced indexing and sends the list to torch.tensor to turn into a tensor, instead of being eligible for `__torch_function__`. I think I might need to hard code a special case for this or something?

Signed-off-by: Edward Yang <ezyang@meta.com>

cc @albanD

### CODE REVIEW

**Comment by ezyang** [comment]
This is ready for review!

File: torch/csrc/functorch/init.cpp [hunk 1]
```diff
@@ -519,6 +526,7 @@ void initFuncTorchBindings(PyObject* module) {
   m.def("is_functionaltensor", &is_functionaltensor);
   m.def("get_unwrapped", &get_unwrapped);
   m.def("maybe_get_level", &maybe_get_level);
+  m.def("_unsafe_set_level", &unsafe_set_level);
```
Comment (albanD) [inline torch/csrc/functorch/init.cpp]: nit: unsafe_maybe_set_level ?

File: torch/_tensor.py [hunk 2]
```diff
@@ -626,6 +626,34 @@ def backward(
             self, gradient, retain_graph, create_graph, inputs=inputs
         )
 
+    def index(self, positions, dims):
```
Comment (albanD) [inline torch/_tensor.py:629]: I would expect this to conflict with this entry from native_functions
```
- func: index.Tensor(Tensor self, Tensor?[] indices) -> Tensor
  device_check: NoCheck   # TensorIterator
  structured_delegate: index.Tensor_out
  variants: function, method
```

But for some reason it doesn't. So I guess we have some logic in the codegen to not actually bind it as a "method".

In the spirit of: avoiding this confusion and not adding a dependency to functorch package here (which I'm trying to remove in another PR as it is just a shell pointing back to torch/_functorch anyways), Can we make this just a function in functorch.dim for now?
Once it is stabilized, I think we should move that code into torch/* and add this method.

File: test/functorch/test_dims.py [hunk 3]
```diff
@@ -106,6 +92,7 @@ def setUp(self):
             self.mem_allocated = torch.cuda.memory_allocated()
 
     def tearDown(self):
+        return
```
Comment (albanD) [inline test/functorch/test_dims.py]: needs to cleanup

File: test/functorch/dim/test_getsetitem.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+        # Test with different tensor
+        tensor2 = torch.randn(2, 3, 4)
+        c, d, e = dims(3)
+        result2 = tensor2[c, d, e]
+        self.assertTrue(isinstance(result2, (torch.Tensor, Tensor)))
+
+    def test_device_handling(self):
+        """Test indexing behavior with different devices."""
+        if torch.cuda.is_available():
```
Comment (albanD) [inline test/functorch/dim/test_getsetitem.py]: nit: onlyCUDA decorator

File: test/functorch/dim/test_getsetitem.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        self.batch, self.height, self.width = dims(3)
+
+    def test_basic_dim_indexing(self):
+        """Test basic indexing with a single Dim."""
+        tensor = torch.randn(3, 4, 5)
+        x, y, z = dims(3)
+
+        # Test indexing with each dim
+        result1 = tensor[x]
+        self.assertTrue(isinstance(result1, (torch.Tensor, Tensor)))
```
Comment (albanD) [inline test/functorch/dim/test_getsetitem.py]: Why are all of these check for both torch.Tensor and dim.Tensor. It should it one or the other? And I guess we are opinionated which one it should be.

File: test/functorch/dim/test_getsetitem.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+        result2 = tensor[y]
+        self.assertTrue(isinstance(result2, (torch.Tensor, Tensor)))
```
Comment (albanD) [inline test/functorch/dim/test_getsetitem.py]: self.assertIsInstance exists

File: functorch/einops/rearrange.py [hunk 7]
```diff
@@ -20,8 +20,6 @@
 
 __all__ = ["rearrange"]
 
-dims = _C.dims
```
Comment (albanD) [inline functorch/einops/rearrange.py:23]: We should mark this as BC-breaking. From my chat with Richard, he was ok with us removing the old code here but we should still mention it on the release notes.

File: functorch/dim/wrap_type.py [hunk 8]
```diff
@@ -26,14 +23,25 @@
...
 
-def wrap_type(to_patch, pattern, __torch_function__):
-    wrap_method = _wrap_method
+def _py_wrap_method(orig: Callable, __torch_function__: Callable) -> Callable:
+    def impl(*args: Any, **kwargs: Any) -> Any:
+        return __torch_function__(orig, None, args, kwargs)
+
+    # Copy metadata from original function
+    impl.__name__ = getattr(orig, "__name__", "")
+    impl.__doc__ = getattr(orig, "__doc__", None)
```
Comment (albanD) [inline functorch/dim/wrap_type.py]: functools.wraps

File: functorch/dim/magic_trace.py [hunk 9]
```diff
@@ -6,11 +6,14 @@
 import os
 import signal
 import subprocess
+from collections.abc import Generator
 from contextlib import contextmanager
 
 
 @contextmanager
-def magic_trace(output="trace.fxt", magic_trace_cache="/tmp/magic-trace"):
+def magic_trace(
```
Comment (albanD) [inline functorch/dim/magic_trace.py:14]: Do you have any doc about this?
Why is this printing stuff? haha

File: functorch/dim/magic_trace.py [hunk 10]
```diff
@@ -26,6 +29,8 @@ def magic_trace(output="trace.fxt", magic_trace_cache="/tmp/magic-trace"):
         subprocess.run(["chmod", "+x", magic_trace_cache])
     args = [magic_trace_cache, "attach", "-pid", str(pid), "-o", output]
     p = subprocess.Popen(args, stderr=subprocess.PIPE, encoding="utf-8")
+    if p.stderr is None:
+        raise RuntimeError("Failed to capture stderr")
```
Comment (albanD) [inline functorch/dim/magic_trace.py]: Is that ever possible?

File: functorch/dim/_wrap.py [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        is_pointwise = False
+
+    wrapper = WrappedOperator(orig, call_torch_function)
+    wrapper.is_pointwise = is_pointwise
+
+    def wrapped_method(self: Any, *args: Any, **kwargs: Any) -> Any:
+        return call_torch_function(wrapper, orig, (type(self),), (self, *args), kwargs)
+
+    # Copy metadata
+    wrapped_method.__name__ = getattr(orig, "__name__", "")
```
Comment (albanD) [inline functorch/dim/_wrap.py]: functool.wraps

File: functorch/dim/_wrap.py [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+    Args:
+        orig: Original method to wrap
+        python_func: Python function wrapper (ignored, we call torch function directly)
+    """
+    # Check if this is a pointwise operation
+    try:
+        from . import op_properties
+
+        is_pointwise = orig in op_properties.pointwise
+    except (ImportError, AttributeError):
```
Comment (albanD) [inline functorch/dim/_wrap.py]: failing to import most likely should mean hard crash?

File: functorch/dim/_wrap.py [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        kwargs = {}
+
+    # Import here to avoid circular imports
+    from . import _Tensor
+
+    # Use the torch function mechanism from _Tensor
+    return _Tensor.__torch_function__(func, types, args, kwargs)
+
+
+def _wrap_method(orig: Callable, python_func: Optional[Callable] = None) -> Callable:
```
Comment (albanD) [inline functorch/dim/_wrap.py]: This is not needed anymore right?

File: functorch/dim/__init__.py [hunk 14]
```diff
@@ -1,13 +1,418 @@
...
+
+
+def dims(
+    n: Optional[int] = None, sizes: Optional[list[Optional[int]]] = None
+) -> Union[Dim, tuple[Dim, ...]]:
+    """
+    Create and return one or more Dim objects.
+
+    Uses bytecode inspection to determine variable names when possible,
+    following the algorithm from functorch/csrc/dim/dim_creation.cpp
```
Comment (albanD) [inline functorch/dim/__init__.py]: old?

File: functorch/dim/__init__.py [hunk 15]
```diff
@@ -1,13 +1,418 @@
...
+    return Dim(name, size if size is not None else -1)
+
+
+def dims(
+    n: Optional[int] = None, sizes: Optional[list[Optional[int]]] = None
+) -> Union[Dim, tuple[Dim, ...]]:
+    """
+    Create and return one or more Dim objects.
+
+    Uses bytecode inspection to determine variable names when possible,
```
Comment (albanD) [inline functorch/dim/__init__.py]: Can we not do the bytecode inspection at all? This still depends on cpython internals that change with versions.

File: functorch/dim/__init__.py [hunk 16]
```diff
@@ -21,80 +426,1159 @@ class DimensionBindError(Exception):
...
+        if hasattr(arg, "__class__") and "torch" in str(type(arg)):
+            safe_args.append(f"<{type(arg).__name__}>")
+        else:
+            safe_args.append(str(arg))
+
+    print(*safe_args, **kwargs, file=sys.stderr)
+
+
+"""
+def _levels_to_tuple(levels: List[DimEntry]) -> tuple[Any, ...]:
```
Comment (albanD) [inline functorch/dim/__init__.py]: Huh?

File: functorch/dim/__init__.py [hunk 17]
```diff
@@ -21,80 +426,1159 @@ class DimensionBindError(Exception):
...
+
     # note that _C.Dim comes before tensor because we want the Dim API for things like size to take precedence.
     # Tensor defines format, but we want to print Dims with special formatting
     __format__ = object.__format__
 
 
-class Tensor(_Tensor, _C.Tensor):
-    from_positional = staticmethod(_C.Tensor_from_positional)
-    sum = _C._instancemethod(_C.Tensor_sum)
+class Tensor(_Tensor):
```
Comment (albanD) [inline functorch/dim/__init__.py:965]: Can we have another name so that type printing is not confusing?

File: functorch/dim/__init__.py [hunk 18]
```diff
@@ -21,80 +426,1159 @@ class DimensionBindError(Exception):
...
+        return result
+
+    @classmethod
+    def create_delayed(
+        cls, orig: Callable, args: tuple, levels: list[DimEntry], has_device: bool
+    ) -> _Tensor:
+        """
+        Create a delayed tensor that defers the operation until later.
+
+        Port of the C++ Tensor::create_delayed method.
```
Comment (albanD) [inline functorch/dim/__init__.py]: We should remove these, they only add confusion for readers

File: functorch/dim/_enable_all_layers.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+import torch
+
+from ._dim_entry import DimEntry
+
+
+if TYPE_CHECKING:
+    from . import Dim, Tensor
+
+
+class EnableAllLayers:
```
Comment (albanD) [inline functorch/dim/_enable_all_layers.py:14]: Why do we need this?

**Review by albanD — COMMENTED** [review]
tbh the code is pretty hard to review with all the comments referencing c++ code that is going to be deleted soon.
Also a lot of code feels like it is not needed? I admit I gave up reviewing some of the new files.

**Comment by ezyang** [comment]
> tbh the code is pretty hard to review with all the comments referencing c++ code that is going to be deleted soon.

To do a proper review, you'll need to pull up C++ code side-by-side

File: torch/_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_tensor.py:629]: I think the main downside is you are temporarily breaking BC (since the FCD index method will stop existing) until you do this. If you like, I can move this code directly to torch/* at the same time of the port.

File: test/functorch/test_dims.py (hunk 3 above)
Comment (ezyang) [reply test/functorch/test_dims.py]: Ah, so probably we should just remove the logic below; it's testing that we aren't generating cycles, but because I've deleted all the non-owning arena logic this may not be so easy to actually achieve in practice.

File: test/functorch/dim/test_getsetitem.py (hunk 4 above)
Comment (ezyang) [reply test/functorch/dim/test_getsetitem.py]: This one and the other one are claude generated extra tests. If you think they're low value we can just delete them.

File: functorch/einops/rearrange.py (hunk 7 above)
Comment (ezyang) [reply functorch/einops/rearrange.py:23]: It's... not BC breaking though? I import dims from functorch.dim so it's still available in this namespace. Unless you're saying that a module import has different assignments than assigning it?

File: functorch/dim/magic_trace.py (hunk 10 above)
Comment (ezyang) [reply functorch/dim/magic_trace.py]: This is probably better as an assert. The coding agent was just trying to get typechecking working lol

File: functorch/dim/magic_trace.py (hunk 9 above)
Comment (ezyang) [reply functorch/dim/magic_trace.py:14]: I actually have no idea what this does lol, the edits here are a side effect of the fact that I turned on typechecking for everything in this folder.

File: functorch/dim/_wrap.py (hunk 13 above)
Comment (ezyang) [reply functorch/dim/_wrap.py]: Yeah, looks like this is unused

File: functorch/dim/__init__.py (hunk 14 above)
Comment (ezyang) [reply functorch/dim/__init__.py]: oopsie, at one point I split dim.cpp into multiple files to make it easier for the agent to retrieve things lol

File: functorch/dim/__init__.py (hunk 15 above)
Comment (ezyang) [reply functorch/dim/__init__.py]: It would be BC breaking to make `x, y = dims()` stop working. I don't think I mind too much dumping it, but we do the tuple destructure bytecode analysis in several other spots in our codebase too so perhaps you will be happy if it gets all unified in one spot.

File: functorch/dim/__init__.py (hunk 16 above)
Comment (ezyang) [reply functorch/dim/__init__.py]: whoops!

File: functorch/dim/__init__.py (hunk 17 above)
Comment (ezyang) [reply functorch/dim/__init__.py:965]: BC breaking! FCD Tensors are called dim.Tensor for some reason. I agree it's confusing and I probably wouldn't have done it for a fresh implementation.

File: functorch/dim/_enable_all_layers.py (hunk 19 above)
Comment (ezyang) [reply functorch/dim/_enable_all_layers.py:14]: This is probably one of the more algorithmically important parts of first class dims. Intuitively, FCD can be thought of as another way of using vmap, where you don't actually have to vmap at the top level, instead the vmaps are implicitly determined by inspecting the bound dimensions on the FCD tensors involved in a compute (this is similar to our concept of non-lexical modes that we spent a long time talking about years ago). But under the hood you still need to actually enable the vmap mode. So once FCD has determined all of the dims we are batching over, it needs to enable all those layers so functorch can actually apply the batching rules. Therefore enable all layers!

File: torch/_tensor.py (hunk 2 above)
Comment (albanD) [reply torch/_tensor.py:629]: Not sure what you mean? These two functions are completely different?

File: functorch/dim/magic_trace.py (hunk 9 above)
Comment (albanD) [reply functorch/dim/magic_trace.py:14]: Ok lol

File: functorch/dim/__init__.py (hunk 15 above)
Comment (albanD) [reply functorch/dim/__init__.py]: From offline chat. That sounds fine to keep for now.

File: test/functorch/test_dims.py (hunk 3 above)
Comment (zdevito) [reply test/functorch/test_dims.py]: wouldn't an API that arbitrarily creates tensor cycles make it a non-starter for actual use? It will fragment memory

**Review by zdevito — APPROVED** [review]
Stamping to unblock, but I am a little concerned about cycles

File: test/functorch/test_dims.py (hunk 3 above)
Comment (ezyang) [reply test/functorch/test_dims.py]: @zdevito hmmm, well, I guess I could introduce weakrefs everywhere we used to have a non-owning reference, which should break the cycles.....

**Review by albanD — APPROVED** [review]
Looks at the last few files, no more comments than what is above.

File: test/functorch/test_dims.py (hunk 3 above)
Comment (ezyang) [reply test/functorch/test_dims.py]: I'm actually not sure how cycles can be created; only dim.Tensor contains DimEntry which points to Dim, it feels like there shouldn't be cycles. I will check this empirically.

File: torch/_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_tensor.py:629]: Look carefully: index is on torch.Tensor! So hypothetically old users of FCD might be relying on being able to call index on plain Tensors to get FCD behavior. Because the method doesn't exist at all you have to add it.

File: test/functorch/test_dims.py (hunk 3 above)
Comment (ezyang) [reply test/functorch/test_dims.py]: It turns out these asserts pass! I am a lucky duck.
